Add authentication and authorization

CrazyTapok opened the pull request:
1. Screenshot: 
![image](https://github.com/user-attachments/assets/44091b0c-3913-46b0-aec6-d9400c014dc1)
![image](https://github.com/user-attachments/assets/3ba37b91-fd68-4a98-8727-8143dd6746ab)



2. What has been implemented: 
  - [x] Add response/request models for login. 
  - [x] Add SessionsController for session/login creation. 
  - [x] Replace the AddEmployeeAsync method with RegisterAsync in the EmployeeController. 
  - [x] Improve reading data from the .env file.  
  - [x] Add IEmployeeRepository and IJwtTokenService. 
  - [x] Implemente interfaces IEmployeeRepository and IJwtTokenService. 
  - [x] Implement GetByEmailAsync method in EmployeeRepository. 
  - [x] Fixed tests EmployeeControllerTests, EmployeeServiceTests and  EmployeeRepositoryTests. 
  - [x] Add SessionsControllerTests. Fixed StartupInfrastructure and ServiceCollectionExtensions.

CrazyTapok committed:
feat: Add authentication and authorization
Add response/request models for login. Add SessionsController for session/login creation. Replace the AddEmployeeAsync method with RegisterAsync in the EmployeeController. Improve reading data from the .env file.  Add IEmployeeRepository and IJwtTokenService. Implemente interfaces IEmployeeRepository and IJwtTokenService. Implement GetByEmailAsync method in EmployeeRepository. Fixed tests EmployeeControllerTests, EmployeeServiceTests and  EmployeeRepositoryTests. Add SessionsControllerTests. Fixed StartupInfrastructure and ServiceCollectionExtensions.

svirzhevskiy commented on src/TaskManager.API/Controllers/EmployeeController.cs:
```diff
@@ -29,11 +31,23 @@ public async Task<ActionResult<EmployeeResponse>> GetEmployeeByIdAsync(Guid id,
+        }
+        
         var employee = request.ToEmployee();
+        employee.Password = PasswordHelper.HashPassword(request.Password);
```
не кажется, что эта логика должна быть в сервисе?

svirzhevskiy replied:
контроллер нужен чтобы хендлить только взаимодействие по вебу, а логику в сервис

svirzhevskiy commented on src/TaskManager.API/Controllers/EmployeeController.cs:
```diff
@@ -29,11 +31,23 @@ public async Task<ActionResult<EmployeeResponse>> GetEmployeeByIdAsync(Guid id,
+            return BadRequest();
+        }
+
+        var existingEmployee = await _employeeService.GetEmployeeByEmailAsync(request.Email, cancellationToken);
+        if (existingEmployee != null)
+        {
+            return Conflict("A user with this email already exists.");
+        }
```
вот это тоже в сервис

svirzhevskiy replied:
чтобы это выглядело просто как вызов сервиса и респонс в зависимости от результата вызова. Ты щас логику сервиса в контролер впихнул

CrazyTapok committed:
refactor: fix remarks after the review
Moved the business logic from the methods of registration/authorization to services. Fix test. Add attributes that restrict access to functionality by roles.

CrazyTapok committed:
fix: remarks
add LoginRequestValidator, add JwtTokenServiceTests, add attributes that restrict access to functionality by roles

svirzhevskiy commented on src/TaskManager.API/Controllers/CompanyController.cs:
```diff
@@ -53,13 +57,15 @@ public async Task<IActionResult> UpdateCompanyAsync(Guid id, [FromBody] CompanyR
     }
 
+    [Authorize(Roles = "Admin,ProjectManager")]
     [HttpGet("companies/{companyId:guid}/employees")]
```
только щас заметил, что у тебя на контроллере висит атрибут [Route("api/[controller]")]
Какой тут роут по итогу? api/companies/companies/... ?

CrazyTapok replied:
![image](https://github.com/user-attachments/assets/a46ac6d8-2603-4ebc-8fed-81d7021f02d9)

svirzhevskiy replied:
так хеня же :)

svirzhevskiy replied:
надо роут api/companies/{id}/employees

svirzhevskiy replied:
в атрибуте на контроллере замени на api/companies, а в аттрибуте метода убери лишний компаниес
И в остальных контроллерах исправь

CrazyTapok replied:
хм, ладно

svirzhevskiy commented on src/TaskManager.API/Controllers/EmployeeController.cs:
```diff
@@ -29,16 +31,33 @@ public async Task<ActionResult<EmployeeResponse>> GetEmployeeByIdAsync(Guid id,
-        var createdEmployee = await _employeeService.AddAsync(employee, cancellationToken);
-        
-        var response = createdEmployee.MapToEmployeeResponse();
+        if (request == null)
+        {
+            return BadRequest("The request cannot be null.");
+        }
```
у тебя есть валидатор для employee request? Если валидатор есть, он отрабатывает до входа в контроллер, те этот код недостижим

svirzhevskiy commented on src/TaskManager.API/Controllers/SessionsController.cs:
```diff
@@ -0,0 +1,32 @@
+    [HttpPost]
+    public async Task<IActionResult> LoginAsync([FromBody] LoginRequest request, CancellationToken cancellationToken = default)
+    {
+        if (request == null)
+        {
+            return BadRequest("Request cannot be null.");
+        }
```
разве этот код может отрабоать при наличии валидатора?

CrazyTapok replied:
я посмотрел, при наличии валидатора, то может оказаться недостижимым

svirzhevskiy replied:
надо удалить значит)

svirzhevskiy commented on src/TaskManager.Core/Services/EmployeeService.cs:
```diff
@@ -1,18 +1,33 @@
 
+    public async Task<Employee?> RegisterAsync(Employee employee, CancellationToken cancellationToken = default)
+    {
+        if (string.IsNullOrWhiteSpace(employee.Email))
+            throw new ArgumentException("Email cannot be null or whitespace.", nameof(employee.Email));
```
у тебя в файле выше однострочный if со скобками. Я бы придерживался одного стиля - либо везде скобки, либо везде без

svirzhevskiy replied:
есть кстати такой метод
`ArgumentNullException.ThrowIfNullOrWhiteSpace(string);`

svirzhevskiy commented on src/TaskManager.Infrastructure/EF/StartupInfrastructure.cs:
```diff
@@ -11,15 +11,18 @@ public static class StartupInfrastructure
     public static void AddDbContextInfrastructure(this IServiceCollection services, IConfiguration configuration)
     {
+        var connectionString = configuration.GetConnectionString("DefaultConnection")
+            ?? throw new InvalidOperationException("Connection string 'DefaultConnection' not found.");
```
мб ArgumentNullException

svirzhevskiy commented on src/TaskManager.Core/Services/JwtTokenService.cs:
```diff
@@ -0,0 +1,38 @@
+
+internal class JwtTokenService(IConfiguration configuration) : IJwtTokenService
+{
+    private readonly IConfiguration _configuration = configuration;
```
ты сюда инжектишь весь конфиг приложения? мб вынести jwt конфиг в отдельный класс и инжектить через IOptions?

CrazyTapok committed:
refactor: fix remarks after the review
fix urls for controllers, removed null checks in employee and session controllers, replace ArgumentException by ArgumentNullException.ThrowIfNullOrWhiteSpace, replace InvalidOperationException by ArgumentNullException in StartupInfrastructure, fix tests, create JwtSettings

CrazyTapok committed:
fix: remarks
remove Null check in employee and session controllers, fix SessionControllerTests
